From c17705196a3f617999f428ec6a0155829baa564f Mon Sep 17 00:00:00 2001
From: Guy Zadicario <guy.zadicario@intel.com>
Date: Sun, 15 Nov 2020 11:12:38 +0200
Subject: [PATCH v11 12/28] misc: nnpi: command channel file interface

Expose an anon file descriptor interface to a command channel object
which allows user-space to send commands to the device by writing to that
file as well as consume device response messages by reading the file.

When the file is released (closed), a channel shut-down sequence starts.
First, a message is sent to the device notifying it that the channel is
closing. Once the response to this message is received from the device,
the command channel object is destroyed.

Signed-off-by: Guy Zadicario <guy.zadicario@intel.com>
Reviewed-by: Vaibhav Agarwal <vaibhav.agarwal@intel.com>
---
 drivers/misc/intel-nnpi/cmd_chan.c | 304 +++++++++++++++++++++++++++++++++++++
 drivers/misc/intel-nnpi/cmd_chan.h |  15 +-
 drivers/misc/intel-nnpi/device.c   | 103 ++++++++++++-
 3 files changed, 418 insertions(+), 4 deletions(-)

diff --git a/drivers/misc/intel-nnpi/cmd_chan.c b/drivers/misc/intel-nnpi/cmd_chan.c
index 8dff915..dc42195 100644
--- a/drivers/misc/intel-nnpi/cmd_chan.c
+++ b/drivers/misc/intel-nnpi/cmd_chan.c
@@ -6,11 +6,13 @@
 
 #include "cmd_chan.h"
 #include <linux/anon_inodes.h>
+#include <linux/bitfield.h>
 #include <linux/dev_printk.h>
 #include <linux/file.h>
 #include <linux/kernel.h>
 #include <linux/slab.h>
 #include "host_chardev.h"
+#include "ipc_c2h_events.h"
 #include "ipc_protocol.h"
 
 #define RESPQ_BUF_SIZE    2048   /* must be power of 2 */
@@ -54,6 +56,213 @@ static inline void respq_pop(struct nnp_chan *chan, char *buf, int count)
 }
 
 /**
+ * respq_is_msg_avail() - check if a response message is available to be popped
+ * @chan: the cmd_chan object
+ *
+ * Checks if new response message is available or channel has been destroyed.
+ *
+ * Return: true if either message is available or channel has destoyed.
+ */
+static inline bool respq_is_msg_avail(struct nnp_chan *chan)
+{
+	bool ret;
+
+	spin_lock_bh(&chan->resp_lock_bh);
+	/*
+	 * response messages are pushed into the respq ring-buffer by pushing
+	 * the size of the message (as u32) followed by message content.
+	 * So an entire message is available only if more than sizeof(u32)
+	 * bytes are available (there is no message with zero size).
+	 */
+	ret = chan->destroyed ||
+	      (CIRC_CNT(chan->respq.head, chan->respq.tail, RESPQ_BUF_SIZE) >
+	       sizeof(u32));
+	spin_unlock_bh(&chan->resp_lock_bh);
+
+	return ret;
+}
+
+static inline int is_cmd_chan_file(struct file *f);
+
+static int cmd_chan_file_release(struct inode *inode, struct file *f)
+{
+	struct nnp_chan *chan = f->private_data;
+	struct file *host_file;
+
+	if (!is_cmd_chan_file(f))
+		return -EINVAL;
+
+	nnp_chan_send_destroy(chan);
+
+	host_file = chan->host_file;
+	nnp_chan_put(chan);
+	fput(host_file);
+
+	return 0;
+}
+
+/**
+ * cmd_chan_file_read() - reads a single response message arrived from device
+ * @f: cmd_chan file descriptor
+ * @buf: buffer to receive the message
+ * @size: size of buf, must be at least 16 qwords (16 * sizeof(u64))
+ * @off: ignored.
+ *
+ * This function will block and wait until interrupted or a response
+ * message from device is available.
+ * When message(s) are available, it reads a single message, copy it to
+ * @buf and returns the message size.
+ * the givven @buf and @size must be large enough to receive the largest
+ * possible message which is 16 qwords, otherwise -EINVAL is returned.
+ *
+ * Return: if positive, the size in bytes of the read message,
+ *         error code otherwise
+ */
+static ssize_t cmd_chan_file_read(struct file *f, char __user *buf, size_t size,
+				  loff_t *off)
+{
+	struct nnp_chan *chan = f->private_data;
+	u32 packet_size;
+	u64 msg[16];  /* maximum possible message in the response queue */
+	int ret;
+
+	if (!is_cmd_chan_file(f))
+		return -EINVAL;
+
+	if (size < sizeof(msg))
+		return -EINVAL;
+
+	/*
+	 * wait for response message to be available, interrupted or channel
+	 * has been destroyed on us.
+	 */
+	ret = wait_event_interruptible(chan->resp_waitq,
+				       respq_is_msg_avail(chan));
+	if (ret < 0)
+		return ret;
+
+	spin_lock_bh(&chan->resp_lock_bh);
+
+	if (chan->destroyed) {
+		spin_unlock_bh(&chan->resp_lock_bh);
+		return -EPIPE;
+	}
+
+	respq_pop(chan, (u8 *)&packet_size, sizeof(u32));
+	/* Check packet_size does not overrun msg size */
+	if (packet_size > sizeof(msg))
+		return -EINVAL;
+	respq_pop(chan, (u8 *)msg, packet_size);
+
+	spin_unlock_bh(&chan->resp_lock_bh);
+
+	ret = copy_to_user(buf, msg, packet_size);
+	if (ret)
+		return -EFAULT;
+
+	return (ssize_t)packet_size;
+}
+
+/**
+ * cmd_chan_file_write() - schedule a command message to be sent to the device.
+ * @f: a cmd_chan file descriptor
+ * @buf: the command message content
+ * @size: size in bytes of the message, must be multiple of 8 and not larger
+ *        than 3 qwords.
+ * @off: ignored
+ *
+ * This function reads a command message from buffer and puts it in the
+ * channel's message queue to schedule it to be delivered to the device.
+ * The function returns when the message is copied to the message scheduler
+ * queue without waiting for it to be sent out.
+ * A valid command message size must be qword aligned and not larger than
+ * the maximum size the message scheduler support, which is 3 qwords.
+ *
+ * The function also validate the command content and fail if the chan_id
+ * field of the command header does not belong to the same channel of this
+ * file descriptor, or the command opcode is out of range, or the command
+ * size does not fit the size of this opcode.
+ *
+ * Return: the size of the message written or error code.
+ */
+static ssize_t cmd_chan_file_write(struct file *f, const char __user *buf,
+				   size_t size, loff_t *off)
+{
+	struct nnp_chan *chan = f->private_data;
+	u64 msg[MSG_SCHED_MAX_MSG_SIZE];
+	unsigned int chan_id, opcode;
+	unsigned int op;
+	int ret;
+
+	if (!is_cmd_chan_file(f))
+		return -EINVAL;
+
+	/*
+	 * size must be multiple of 8 bytes and cannot exceed maximum message
+	 * size
+	 */
+	if ((size > MSG_SCHED_MAX_MSG_SIZE * 8) ||
+	    (size &  0x7) != 0)
+		return -EINVAL;
+
+	ret = copy_from_user(msg, buf, size);
+	if (ret != 0)
+		return -EFAULT;
+
+	/*
+	 * Check chan_id, opcode and message size are valid
+	 */
+	opcode = FIELD_GET(NNP_H2C_CHAN_MSG_OP_MASK, msg[0]);
+	chan_id = FIELD_GET(NNP_H2C_CHAN_MSG_CHAN_ID_MASK, msg[0]);
+	if (chan_id != chan->chan_id)
+		return -EINVAL;
+	if (opcode < USER_MIN_OPCODE)
+		return -EINVAL;
+	op = opcode - USER_MIN_OPCODE;
+	if (size != (chan->nnpdev->ipc_chan_cmd_op_size[op] * 8))
+		return -EINVAL;
+
+	if (!is_card_fatal_drv_event(chan_broken(chan)))
+		ret  = nnp_msched_queue_add_msg(chan->cmdq, msg, size / 8);
+	else
+		ret = -EPIPE;
+
+	if (ret < 0)
+		return ret;
+	else
+		return size;
+}
+
+static unsigned int cmd_chan_file_poll(struct file *f,
+				       struct poll_table_struct *pt)
+{
+	struct nnp_chan *chan = f->private_data;
+	unsigned int mask = (POLLOUT | POLLWRNORM);
+
+	if (!is_cmd_chan_file(f))
+		return 0;
+
+	poll_wait(f, &chan->resp_waitq, pt);
+	if (respq_is_msg_avail(chan))
+		mask |= (POLLIN | POLLRDNORM);
+
+	return mask;
+}
+
+static const struct file_operations nnp_chan_fops = {
+	.owner = THIS_MODULE,
+	.release = cmd_chan_file_release,
+	.read = cmd_chan_file_read,
+	.write = cmd_chan_file_write,
+	.poll = cmd_chan_file_poll,
+};
+
+static inline int is_cmd_chan_file(struct file *f)
+{
+	return f->f_op == &nnp_chan_fops;
+}
+
+/**
  * nnpdev_chan_create() - creates a command channel object
  * @nnpdev: the device
  * @host_fd: opened file descriptor to "/dev/nnpi_host"
@@ -117,6 +326,7 @@ int nnpdev_chan_create(struct nnp_device *nnpdev, int host_fd,
 	cmd_chan->chan_id = chan_id;
 	nnpdev_get(nnpdev);
 	cmd_chan->nnpdev = nnpdev;
+	cmd_chan->fd = -1;
 	cmd_chan->get_device_events = get_device_events;
 
 	cmd_chan->nnp_user = cmd_chan->host_file->private_data;
@@ -166,6 +376,9 @@ static void nnp_chan_release(struct kref *kref)
 	ida_simple_remove(&cmd_chan->nnpdev->cmd_chan_ida,
 			  cmd_chan->chan_id);
 
+	if (cmd_chan->fd < 0)
+		fput(cmd_chan->host_file);
+
 	nnp_user_put(cmd_chan->nnp_user);
 
 	kfree(cmd_chan->respq_buf);
@@ -184,6 +397,97 @@ void nnp_chan_put(struct nnp_chan *cmd_chan)
 	kref_put(&cmd_chan->ref, nnp_chan_release);
 }
 
+int nnp_chan_create_file(struct nnp_chan *cmd_chan)
+{
+	/*
+	 * get refcount to the channel that will drop when
+	 * the file is released.
+	 */
+	nnp_chan_get(cmd_chan);
+
+	cmd_chan->fd = anon_inode_getfd("nnpi_chan", &nnp_chan_fops,
+					cmd_chan, O_RDWR | O_CLOEXEC);
+	if (cmd_chan->fd < 0)
+		nnp_chan_put(cmd_chan);
+
+	return cmd_chan->fd;
+}
+
+/**
+ * nnp_chan_set_destroyed() - atomically mark the channel "destroyed"
+ * @chan: the cmd_chan
+ *
+ * This function sets the command channel state to "destroyed" and returns
+ * the previous destroyed state.
+ * This function should be called once the channel has been destructed on the
+ * device and a "channel destroyed" response message arrived.
+ *
+ * Return: true if the channel was already marked destroyed.
+ */
+bool nnp_chan_set_destroyed(struct nnp_chan *chan)
+{
+	bool ret;
+
+	spin_lock(&chan->resp_lock_bh);
+	ret = chan->destroyed;
+	chan->destroyed = true;
+	spin_unlock(&chan->resp_lock_bh);
+
+	wake_up_all(&chan->resp_waitq);
+
+	return ret;
+}
+
+/**
+ * nnp_chan_send_destroy() - sends a "destroy channel" command to device
+ * @chan: the cmd_chan to destroy.
+ *
+ * This function sends a command to the device to destroy a command channel,
+ * The channel object remains to exist, it will be dropped only when the device
+ * send back a "channel destroyed" response message, or when the device is in
+ * critical error where we treat it as "not function".
+ *
+ * Return: 0 on success, error value otherwise.
+ */
+int nnp_chan_send_destroy(struct nnp_chan *chan)
+{
+	u64 cmd;
+	int ret = 0;
+	bool do_put = false;
+
+	spin_lock(&chan->resp_lock_bh);
+	if (chan->destroyed)
+		goto done;
+
+	cmd = FIELD_PREP(NNP_H2C_OP_MASK, NNP_IPC_H2C_OP_CHANNEL_OP);
+	cmd |= FIELD_PREP(NNP_H2C_CHANNEL_OP_CHAN_ID_MASK, chan->chan_id);
+	cmd |= FIELD_PREP(NNP_H2C_CHANNEL_OP_DESTROY_MASK, 1);
+
+	chan->event_msg.value = 0;
+
+	/*
+	 * If card is in critical state (or was during the channel lifetime)
+	 * we destroy the channel.
+	 * otherwise, we send a destroy command to card and will destroy when
+	 * the destroy reply arrives.
+	 */
+	if (is_card_fatal_drv_event(chan_broken(chan))) {
+		chan->destroyed = true;
+		do_put = true;
+		goto done;
+	}
+
+	ret = nnpdev_queue_msg(chan->nnpdev->cmdq, cmd);
+
+done:
+	spin_unlock(&chan->resp_lock_bh);
+	if (do_put) {
+		wake_up_all(&chan->resp_waitq);
+		nnp_chan_put(chan);
+	}
+	return ret;
+}
+
 /**
  * nnp_chan_add_response() - adds a response message targeting this channel
  * @cmd_chan: the command channel object
diff --git a/drivers/misc/intel-nnpi/cmd_chan.h b/drivers/misc/intel-nnpi/cmd_chan.h
index 87758d4..64346df 100644
--- a/drivers/misc/intel-nnpi/cmd_chan.h
+++ b/drivers/misc/intel-nnpi/cmd_chan.h
@@ -24,9 +24,11 @@
  * @chan_id: the ipc channel id for this channel
  * @hash_node: node to include this object in list of channels
  *             hash is in (cmd_chan_hash in nnp_device).
+ * @event_msg: ipc event response received from device during create channel
  * @card_critical_error: last critical event report received from device
  * @get_device_events: true if device-level events received from device should
  *                     be sent over this channel to user.
+ * @fd: file descriptor created for the channel (implements read/write)
  * @cmdq: message queue added to msg_scheduler, for user commands to be sent
  *        to the device.
  * @host_file: reference to opened "/dev/nnpi_host" object which defines the
@@ -34,7 +36,9 @@
  * @nnp_user: the nnp_user this channel belongs to.
  *             the channel can reference host resources created by this
  *             nnp_user object.
- * @resp_lock_bh: protects @respq
+ * @resp_lock_bh: protects accesses to @respq and @destroyed
+ * @destroyed: a state indicating that the channel should be not-yet-exist on
+ *             the device.
  * @resp_waitq: waitqueue used for waiting for response messages be available.
  * @respq: circular buffer object that receive response messages from device.
  * @respq_buf: buffer space allocated for circular response buffer.
@@ -44,16 +48,19 @@ struct nnp_chan {
 	struct nnp_device      *nnpdev;
 	u16                    chan_id;
 	struct hlist_node      hash_node;
+	union c2h_event_report event_msg;
 	union c2h_event_report card_critical_error;
 	bool                   get_device_events;
 
+	int fd;
 	struct nnp_msched_queue     *cmdq;
 	struct file                *host_file;
 	struct nnp_user_info       *nnp_user;
 
 	wait_queue_head_t resp_waitq;
+	bool              destroyed;
 
-	spinlock_t        resp_lock_bh; /* protects respq modifications */
+	spinlock_t        resp_lock_bh; /* protects @respq, @destroyed */
 	struct circ_buf   respq;
 	char             *respq_buf;
 };
@@ -67,6 +74,10 @@ int nnpdev_chan_create(struct nnp_device *nnpdev, int host_fd,
 void nnp_chan_get(struct nnp_chan *cmd_chan);
 void nnp_chan_put(struct nnp_chan *cmd_chan);
 
+int nnp_chan_create_file(struct nnp_chan *cmd_chan);
+int nnp_chan_send_destroy(struct nnp_chan *chan);
+bool nnp_chan_set_destroyed(struct nnp_chan *chan);
+
 int nnp_chan_add_response(struct nnp_chan *cmd_chan, u64 *hw_msg, u32 size);
 
 #endif
diff --git a/drivers/misc/intel-nnpi/device.c b/drivers/misc/intel-nnpi/device.c
index 25ad756..91bf1ff 100644
--- a/drivers/misc/intel-nnpi/device.c
+++ b/drivers/misc/intel-nnpi/device.c
@@ -228,6 +228,48 @@ struct nnp_chan *nnpdev_find_channel(struct nnp_device *nnpdev, u16 chan_id)
 	return NULL;
 }
 
+/**
+ * set_destroyed_one_channel() - marks one channel as destroyed.
+ * @nnpdev: The nnp device
+ *
+ * The function search for a channel that is not yet marked as destroyed,
+ * mark it as destriyed and returns a pointer to the channel.
+ *
+ * Return: NULL if no channel found which did not yet marked destroyed,
+ * otherwise a pointer to the newly marked destroyed channel.
+ */
+static struct nnp_chan *set_destroyed_one_channel(struct nnp_device *nnpdev)
+{
+	struct nnp_chan *cmd_chan;
+	int i;
+
+	spin_lock(&nnpdev->lock);
+	hash_for_each(nnpdev->cmd_chan_hash, i, cmd_chan, hash_node) {
+		if (!nnp_chan_set_destroyed(cmd_chan)) {
+			spin_unlock(&nnpdev->lock);
+			return cmd_chan;
+		}
+	}
+	spin_unlock(&nnpdev->lock);
+
+	return NULL;
+}
+
+static void destroy_all_channels(struct nnp_device *nnpdev)
+{
+	struct nnp_chan *cmd_chan;
+
+	/*
+	 * re-iterate on all channel hash until it does not include any
+	 * channel which does not yet marked destroyed.
+	 */
+	cmd_chan = set_destroyed_one_channel(nnpdev);
+	while (cmd_chan) {
+		nnp_chan_put(cmd_chan);
+		cmd_chan = set_destroyed_one_channel(nnpdev);
+	}
+}
+
 void nnpdev_submit_device_event_to_channels(struct nnp_device *nnpdev,
 					    union c2h_event_report *event_msg,
 					    bool force)
@@ -248,6 +290,45 @@ void nnpdev_submit_device_event_to_channels(struct nnp_device *nnpdev,
 					      sizeof(*event_msg));
 	}
 	spin_unlock(&nnpdev->lock);
+
+	/*
+	 * On card fatal event, we consider the device dead and there is
+	 * no point communicating with it. The user will destroy the channel
+	 * and initiate a device reset to fix this.
+	 * We destroy all channels here since the NNP_IPC_CHANNEL_DESTROYED
+	 * response, which normally do that, will never arrive.
+	 */
+	if (is_card_fatal_drv_event(event_msg->event_code))
+		destroy_all_channels(nnpdev);
+}
+
+static void handle_channel_destroy(struct nnp_device *nnpdev,
+				   union c2h_event_report *event_msg)
+{
+	struct nnp_chan *cmd_chan;
+
+	cmd_chan = nnpdev_find_channel(nnpdev, event_msg->obj_id);
+	if (!cmd_chan) {
+		dev_err(nnpdev->hw_dev->dev,
+			"Got channel destroyed reply for not existing channel %d\n",
+			event_msg->obj_id);
+		return;
+	}
+
+	/*
+	 * channel is destroyed on device. Put the main ref of cmd_chan if it
+	 * did not already done.
+	 * There is one possible case that the channel will be already marked
+	 * as destroyed when we get here. This is when we got some card fatal
+	 * event, which caused us to flag the channel as destoyed, but later
+	 * the "destroy channel" response has arrived from the device
+	 * (unexpected).
+	 */
+	if (!nnp_chan_set_destroyed(cmd_chan))
+		nnp_chan_put(cmd_chan);
+
+	/* put against the get from find_channel */
+	nnp_chan_put(cmd_chan);
 }
 
 /*
@@ -257,9 +338,27 @@ void nnpdev_submit_device_event_to_channels(struct nnp_device *nnpdev,
 static void process_device_event(struct nnp_device *nnpdev,
 				 union c2h_event_report *event_msg)
 {
-	bool force = is_card_fatal_event(event_msg->event_code);
+	if (is_card_fatal_event(event_msg->event_code)) {
+		nnpdev_submit_device_event_to_channels(nnpdev, event_msg, true);
+	} else {
+		switch (event_msg->event_code) {
+		case NNP_IPC_DESTROY_CHANNEL_FAILED:
+			dev_err(nnpdev->hw_dev->dev,
+				"Channel destroyed failed channel %d val %d\n",
+				event_msg->obj_id, event_msg->event_val);
+			fallthrough;
+		case NNP_IPC_CHANNEL_DESTROYED:
+			handle_channel_destroy(nnpdev, event_msg);
+			break;
+		default:
+			dev_err(nnpdev->hw_dev->dev,
+				"Unknown event received - %u\n",
+				event_msg->event_code);
+			return;
+		}
 
-	nnpdev_submit_device_event_to_channels(nnpdev, event_msg, force);
+		nnpdev_submit_device_event_to_channels(nnpdev, event_msg, false);
+	}
 }
 
 struct event_report_work {
-- 
1.8.3.1

